assets/tests/state.test.ts

```diff
@@ -769,4 +792,101 @@ describe("reducer", () => {
       routeTabsPushInProgress: false,
     })
   })
+
+  test("closeInputModal", () => {
+    const mockCallback = jest.fn()
+    const stateWithInputModal = {
+      ...initialState,
+      openInputModal: {
+        type: "DELETE_PRESET" as "DELETE_PRESET",
+        deleteCallback: mockCallback,
+        presetName: "My Preset",
+      },
+    }
+
+    const newState = reducer(stateWithInputModal, State.closeInputModal())
+
+    expect(newState.openInputModal).toBeNull()
+  })
+
+  test("promptToSaveOrCreatePreset when creating", () => {
+    const mockDispatch = jest.fn()
+    const routeTab = routeTabFactory.build({ presetName: undefined })
+
+    const newState = reducer(
+      initialState,
+      State.promptToSaveOrCreatePreset(routeTab)
+    )
+
+    switch (newState.openInputModal?.type) {
+      case "CREATE_PRESET":
+        newState.openInputModal.createCallback("Preset name", mockDispatch)
+        expect(mockDispatch).toHaveBeenCalledWith(
+          State.createPreset(routeTab.uuid, "Preset name")
+        )
+        return
+      default:
+        fail("did not receive correct openInputModal type")
```

Comment:
The reason is so that on line 823 the typechecker realizes that `newState.openInputModal.createCallback` is a thing without me having to do any manual `as` annotations, and I personally like avoiding those where possible because explicit `as` annotations can hide incorrect judgements about what the type of something should be.

Reply:
another reason to use data instead of callbacks in the state 😄 . but I think it'd be fine to do `state.saveCallback!()`, since it will crash if it goes wrong, so the test will still catch any issues. the `fail` call is a little more explicit and would give a better message, but the cost to readability might not be worth it.

Reply:
Well the different types of input modals still have different attributes, so just because they're not functions doesn't inherently fix that. I'd like this to be general enough that it can expand to modals prompting users for things totally unrelated to presets.

Also FWIW `saveCallback!` doesn't actually work here since the issue isn't that `saveCallback` might exist as a key but be undefined, it's that it's not present on some types of input modals, so the typechecker still complains (unless you additionally throw an `as` in there).